src/entities/join/composable-stable/composableStableJoin.ts

```diff
@@ -0,0 +1,215 @@
+            poolId: poolState.id,
+            sender: ZERO_ADDRESS,
+            recipient: ZERO_ADDRESS,
+            maxAmountsIn: amounts.maxAmountsIn,
+            userData,
+            fromInternalBalance: input.fromInternalBalance ?? false,
+        });
+
+        const queryResult = await doQueryJoin(
+            input.rpcUrl,
+            input.chainId,
+            args,
+        );
+
+        const bpt = new Token(input.chainId, poolState.address, 18);
+        const bptOut = TokenAmount.fromRawAmount(bpt, queryResult.bptOut);
+
+        const amountsIn = queryResult.amountsIn.map((a, i) =>
+            TokenAmount.fromRawAmount(tokensIn[i], a),
+        );
+
+        return {
+            poolType: poolState.type,
+            joinKind: input.kind,
+            poolId: poolState.id,
+            bptOut,
+            amountsIn,
+            tokenInIndex: amounts.tokenInIndex,
+            fromInternalBalance: !!input.fromInternalBalance,
+            bptIndex,
+        };
+    }
+
+    public buildCall(input: ComposableJoinCall): JoinBuildOutput {
+        const amounts = this.getAmountsCall(input);
+        const amountsWithoutBpt = {
+            ...amounts,
+            maxAmountsIn: [
+                ...amounts.maxAmountsIn.slice(0, input.bptIndex),
+                ...amounts.maxAmountsIn.slice(input.bptIndex + 1),
+            ],
+        };
+
+        const userData = this.encodeUserData(input.joinKind, amountsWithoutBpt);
+
+        const { args } = parseJoinArgs({
+            ...input,
+            sortedTokens: input.amountsIn.map((a) => a.token),
+            maxAmountsIn: amounts.maxAmountsIn,
+            userData,
+            fromInternalBalance: input.fromInternalBalance,
+        });
+
+        const call = encodeFunctionData({
+            abi: vaultAbi,
+            functionName: 'joinPool',
+            args,
+        });
+
+        const value = input.amountsIn.find(
+            (a) => a.token.address === ZERO_ADDRESS,
+        )?.amount;
+
+        return {
+            call,
+            to: BALANCER_VAULT,
+            value,
+            minBptOut: amounts.minimumBpt,
+            maxAmountsIn: amounts.maxAmountsIn,
+        };
+    }
+
+    private getAmountsQuery(
+        poolTokens: Token[],
+        input: JoinInput,
+        bptIndex?: number,
+    ): AmountsJoin {
+        switch (input.kind) {
+            case JoinKind.Init:
+            case JoinKind.Unbalanced: {
+                return {
+                    minimumBpt: 0n,
+                    maxAmountsIn: getAmounts(
+                        poolTokens,
+                        input.amountsIn,
+                        BigInt(0),
+                    ),
+                    tokenInIndex: undefined,
+                };
+            }
+            case JoinKind.SingleAsset: {
+                if (bptIndex === undefined) {
+                    throw new Error('bptIndex is necessary');
+                }
```

Comment:
I feel like these checks are better suited for a `validateInputs` method that should run before anything else. What do you think?

Reply:
Actually, this bptIndex will not be undefined in any case, this "if" was necessary before I adapted the types, I will remove this, but thanks

Reply:
You'll still have to check for bptIndex being different than -1 somewhere, right?

Reply:
Right, that's true